File: SQM/SoftwareAnalyzer2/Structure/Metrics/QAlgorithm.cs
using SoftwareAnalyzer2.Structure.Gephi;
using SoftwareAnalyzer2.Structure.Graphing;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SoftwareAnalyzer2.Structure.Metrics
{
    class QAlgorithm
    {
        private static QAlgorithm only;
        private QAlgorithm()
        {
        }

        internal static QAlgorithm Instance
        {
            get
            {
                if (only == null)
                {
                    only = new QAlgorithm();
                }
                return only;
            }
        }

        internal void MainAlgorithm()
        {
            //The original algorithm had a minimum number of modules which would be acceptable as a solution.
            //This assumes some known information about the graph.
            double minDiff = .00001;//this value defines the minimum gain between iterations

            double q1 = CalcModularityQ();//get initial modularity
            double q0 = q1 + 1;//force the first iteration
            int iteration = 0;

            bool GraphRevised = false;
            //iterate until no significant difference is acheived
            //the old algorithm included an escape clause here which depended on the minimum nodes 
            while (Math.Abs(q0 - q1) > minDiff || GraphRevised)
            {
                GraphRevised = false;
                q0 = q1;//remember the last measure of modularity
                GraphRevised = CombineModules();//attempt to combine modules
                q1 = CalcModularityQ();//find new modularity
                iteration++;
            }
        }

        private double CalcModularityQ()
        {
            double mValue = 0;
            double sum = 0;
            long module1 = 0;
            long module2 = 0;

            foreach (GephiNode first in MetricUtilities.AllNodes)
            {
                //this node has at least one connection
                module1 = (long)first.GetProperty(NodeProperties.Module);
                foreach (GephiNode second in MetricUtilities.GetDirectedLinksFrom(first))
                {
                    if (second == first)
                    {
                        throw new InvalidDataException("No self-referential links allowed in Q algorithm");
                    }

                    module2 = (long)second.GetProperty(NodeProperties.Module);
                    double qVal = MetricUtilities.GetDirectedLink(first, second);
                    mValue += qVal;
                    if (module2 == module1)//this is an inter-module connection
                    {
                        sum += qVal;//count this weight.
                    }
                }
            }
            return sum / mValue;
        }

        /// <summary>
        /// Examines the benefit in terms of Q from joining n1 to n2.
        /// Returns a positive number if there is an advantage
        /// Returns a negative number if the joining should be reversed (i.e. n2 --> n1)
        /// Returns 0 if there is no advantage in either direction
        /// </summary>
        /// <param name="firstNode"></param>
        /// <param name="secondNode"></param>
        /// <param name="deltaQ"></param>
        /// <returns></returns>
        private int CalcShortQ(GephiNode firstNode, GephiNode secondNode, out double deltaQ)
        {
            long firstModule = (long)firstNode.GetProperty(NodeProperties.Module);
            long secondModule = (long)secondNode.GetProperty(NodeProperties.Module);

            //we need to find the value lost from links between n1 and its co-module members
            double firstModuleDelta = 0;
            foreach (GephiNode firstModuleNode in MetricUtilities.GetModule(firstModule))
            {
                //skip the same node
                if (firstModuleNode == firstNode)
                {
                    continue;
                }

                //update the firstModuleDelta negatively because we LOSE these links
                firstModuleDelta -= MetricUtilities.GetDirectedLink(firstNode,firstModuleNode);
                firstModuleDelta -= MetricUtilities.GetDirectedLink(firstModuleNode, firstNode);
            }

            //we need to find the value gained from links between n1 and the new module
            foreach (GephiNode secondModuleNode in MetricUtilities.GetModule(secondModule))
            {
                //skip any second module node which doesn't link to the first node
                if (!MetricUtilities.IsThereALinkBetween(firstNode,secondModuleNode))
                {
                    continue;
                }

                //update the firstModuleDelta positively because we GAIN these links
                firstModuleDelta += MetricUtilities.GetDirectedLink(firstNode, secondModuleNode);
                firstModuleDelta += MetricUtilities.GetDirectedLink(secondModuleNode, firstNode);
            }

            //we need to find the value lost from links between n2 and its co-module members
            double secondModuleDelta = 0;
            foreach (GephiNode secondModuleNode in MetricUtilities.GetModule(secondModule))
            {
                //skip the same node
                if (secondModuleNode == secondNode)
                {
                    continue;
                }

                //update the secondModuleDelta negatively because we LOSE these links
                secondModuleDelta -= MetricUtilities.GetDirectedLink(secondNode,secondModuleNode);
                secondModuleDelta -= MetricUtilities.GetDirectedLink(secondModuleNode, secondNode);
            }

            //we need to find the value gained from links between n2 and the new module
            foreach (GephiNode firstModuleNode in MetricUtilities.GetModule(firstModule))
            {
                //skip any first module node which doesn't link to the second node
                if (!MetricUtilities.IsThereALinkBetween(secondNode,firstModuleNode))
                {
                    continue;
                }

                //update the secondModuleDelta positively because we will GAIN these links
                secondModuleDelta += MetricUtilities.GetDirectedLink(secondNode,firstModuleNode);
                secondModuleDelta += MetricUtilities.GetDirectedLink(firstModuleNode, secondNode);
            }

            //determine the best answer
            int direction = 0;
            deltaQ = 0;
            if (firstModuleDelta >= secondModuleDelta && firstModuleDelta > 0)//favor answer1 if it is superior or equal
            {
                deltaQ = firstModuleDelta;
                direction = 1;
            }
            else if (secondModuleDelta > firstModuleDelta && secondModuleDelta > 0)
            {
                deltaQ = secondModuleDelta;
                direction = -1;
            }

            return direction;
        }

        private bool CombineModules()
        {
            bool answer = false;

            #region Move single node
            //consider moving each node in the QNode set
            foreach (GephiNode mover in MetricUtilities.AllNodes)
            {
                AbbreviatedGraph ag = MetricUtilities.GraphFromGephi(mover);
                if (!((bool)mover.GetProperty(NodeProperties.QState) && MetricUtilities.IsImportantConnection(mover)))
                {
                    continue;// no outgoing edges from this node
                }
                
                //configure answer values which indicate the most advantageous movement
                //default values indicate no change is preferrable.
                double bestDeltaQ = 0;
                GephiNode bestLink = mover;
                int change = 0;

                long moverModuleID = (long)mover.GetProperty(NodeProperties.Module);//saved to allow temporary changes with reversion

                foreach (GephiNode linked in MetricUtilities.GetDirectedLinksFrom(mover))
                {
                    long linkedModule = (long)linked.GetProperty(NodeProperties.Module);
                    bool inQAlgorithm = (bool)linked.GetProperty(NodeProperties.QState);
                    if (linkedModule == moverModuleID || !inQAlgorithm)
                    {
                        //we do not consider linked nodes in the same module our outside the q algorithm
                        continue;
                    }

                    //find the best value of the proposed change
                    double deltaQ;
                    int direction = CalcShortQ(mover, linked, out deltaQ);

                    //continue if the value is not more advantageous
                    if (deltaQ <= bestDeltaQ)
                    {
                        continue;
                    }

                    //recall the direction and value of the proposed change
                    change = direction;
                    bestLink = linked;
                    bestDeltaQ = deltaQ;
                }

                //effect the best change
                long otherModuleID = (long)bestLink.GetProperty(NodeProperties.Module);

                if (change == 1)//mover goes to other module
                {
                    MetricUtilities.ChangeModule(mover, otherModuleID);
                    answer = true;//remember that the graph has been revised.
                }
                else if (change == -1) // moving some node to g's module
                {
                    MetricUtilities.ChangeModule(bestLink, moverModuleID);
                    answer = true;//remember that the graph has been revised.
                }
            }
            #endregion

            //recall any module which is well separated
            List<long> separatedModules = new List<long>();

            //consider any super modules
            long myModule = -1;
            long otherModule = -1;
            bool allowSuperModule = true;
            foreach (long mod in MetricUtilities.ModuleIds)
            {
                //reset search parameters
                myModule = mod;
                otherModule = -1;
                allowSuperModule = true;
                bool someValidQ = false;

                //look at each node in the module
                if (MetricUtilities.ModuleExists(mod))
                {
                    //this module has not been removed by creating a super module
                    foreach (GephiNode n in MetricUtilities.GetModule(mod))
                    {
                        AbbreviatedGraph ag = MetricUtilities.GraphFromGephi(n);

                        if ((bool)n.GetProperty(NodeProperties.QState))
                        {
                            someValidQ = true;

                            //examine connections to other modules
                            foreach (GephiNode g2 in MetricUtilities.GetLinksInvolving(n))
                            {
                                long otherMod = (long)g2.GetProperty(NodeProperties.Module);
                                bool inQAlgorithm = (bool)g2.GetProperty(NodeProperties.QState);
                                if (otherMod == myModule || !inQAlgorithm)
                                {
                                    //we do not consider linked nodes in the same or external modules
                                    continue;
                                }

                                if (otherModule != -1 && otherMod != otherModule)
                                {
                                    //we already have a connection to one module
                                    //but this node is from a different module
                                    //therefore they cannot become a super module
                                    allowSuperModule = false;
                                    break;//leave one for loop
                                }
                                else if (otherModule == -1)
                                {
                                    //we have not found any external module
                                    //so we will recall the potential for linking a module.
                                    otherModule = otherMod;
                                }
                            }

                            if (!allowSuperModule)
                            {
                                //we have already established that at least two other modules touch this one.
                                break;//preempt rest of module loop
                            }
                        }
                    }
                }

                //finished examining members of one module

                if (!someValidQ)
                {
                    //this module has no q members, so we won't move it
                    allowSuperModule = false;
                }
                else if (otherModule == -1)
                {
                    //module was not connected to any other module and at least one node was still in the q state
                    separatedModules.Add(myModule);
                    allowSuperModule = false;
                }

                if (allowSuperModule)
                {
                    //we found one super module; combine the two modules
                    answer = true;//remember that the graph has been revised.

                    //remove the other module from the list of separatedModules as it might have been put there.
                    separatedModules.Remove(otherModule);

                    GephiNode[] otherList = new GephiNode[MetricUtilities.SizeOfModule(otherModule)];

                    foreach (GephiNode n in MetricUtilities.GetModule(otherModule))
                    {
                        MetricUtilities.ChangeModule(n, myModule);
                    }
                }
            }

            //create any separated modules
            foreach (long moduleID in separatedModules)
            {
                answer = true;//remember that the graph has been revised.

                //remove those nodes from the Q algorithm and tag them as external
                foreach (GephiNode n in MetricUtilities.GetModule(moduleID))
                {
                    n.SetProperty(NodeProperties.QState, false);
                    n.SetProperty(NodeProperties.NodePattern, (int)NodePatterns.Separated);
                }
            }

            return answer;
        }
    }
}

File: ModularityRCode/Qalgorithm.R
library(tidyverse)
library(igraph)
library(tictoc)

# runs everything
mainAlgorithm <- function() {
    # minimum gain needed to continue iterating 
    minDiff <- 0.00001 
    
    # get initial modularity
    q1 <- CalcModularityQ()
    if(debug) {
        print(paste0("initial mod: ", q1))   
    }
    # forces first iteration (in while loop q0 - q1 will be greater than minDiff)
    q0 <- q1 + 1 
    iteration <- 0
    
    # another way to keep track of continuation
    graphRevised <- FALSE
    
    # iterate until no modularity differences occur
    while(abs(q0 - q1) > minDiff | graphRevised) {
        iteration <- iteration + 1
        graphRevised <- FALSE
        # update last measure of modularity
        q0 <- q1
        # attempt to combine modules 
        graphRevised = combineModules(FALSE)
        # recalculate modularity
        q1 <- CalcModularityQ()
        if(debug){
            print(paste0("q1: ", q1))
            print(paste0("iteration: ", iteration))
            if (iteration == 4) {
              return(iteration)
            }   
        }
    }
    print(paste0("final Modularity: ", q1))
    print(paste0("final modules: ", 1))
    answer <- finalModules()
    anTrim <- answer[ answer != 0 ]
    print(anTrim)
    
}

# actual calculation of Q value 
finalModules <- function() {
    # empty list of size Modules
    modList <- integer(length(modules))
    print(modList)
        
    # cycle through all node_list NOT VERY EFFICIENT 
    for (first in node_list$V1) {
        # get the first node_list module
        mod1 <- node_list[first, 2]
        
        # get all linked node_list to current node using adjacency matrix
        lN_list <- which(AM[first, ] != 0, arr.ind = FALSE)
        
        # check modularity with all node_list 
        for (second in lN_list) {
            # don't want to look at same node
            if (second != first) {
                # get the second node_list module
                mod2 <- node_list[second, 2]
                # value of 1 new edge we look for (weight depends on module connection?)
                qVal <- AM[first, second]
                # inter module connection 
                if (mod2 == mod1){
                    # add to counter of edges within a module
                    modList[[mod1]] <- modList[[mod1]] + c(qVal)
                }
            }
            
        }
    }
  
    return(modList)
}

# actual calculation of Q value 
CalcModularityQ <- function() {
    # initialize variables
    mValue <- 0
    sum <- 0
    
    # cycle through all node_list NOT VERY EFFICIENT 
    for (first in node_list$V1) {
        # get the first node_list module
        mod1 <- node_list[first, 2]
        
        # get all linked node_list to current node using adjacency matrix
        lN_list <- which(AM[first, ] != 0, arr.ind = FALSE)
        
        # check modularity with all node_list 
        for (second in lN_list) {
            # don't want to look at same node
            if (second != first) {
                # get the second node_list module
                mod2 <- node_list[second, 2]
                # value of 1 new edge we look for (weight depends on module connection?)
                qVal <- AM[first, second]
                # total num of edges that exist
                mValue <- mValue + qVal
                # inter module connection 
                if (mod2 == mod1){
                    # add to counter of edges within a module
                    sum <- sum + qVal 
                }
            }
            
        }
        node_list <<- node_list
    }
    # ratio of edges inside of modules to total edges 
    return(as.double(sum/mValue))
}

# myopic view of Q value   
# returns positive num if advantage, negative num if joining should be reversed, 0 if no advantage
CalcShortQ <- function(firstNode, secondNode) {
    # get number of the module each node is connected to
    firstModule <<- node_list[firstNode, 2]
    secondModule <- node_list[secondNode, 2]
    
    firstModuleDelta <- 0
    secondModuleDelta <- 0 
    
    if (debug) {
        print(paste0("first group: ", modules[[firstModule]]))
    }
    if (length(modules[[firstModule]]) > 0) {
        for (firstModNode in modules[[firstModule]]) {
            # update for lost links when we move module n1 to new module
            # skip the same node 
            if (firstModNode != firstNode) {
                # update the firstModuleDelta negatively because links are lost
                # get directed link through an adjacency matrix
                #print(c(-1, firstNode, firstModNode, -AM[firstModNode,firstNode], -AM[firstNode,firstModNode]))
                firstModuleDelta <- firstModuleDelta - AM[firstModNode, firstNode] - AM[firstNode,firstModNode]
            }
            # update the value gained from links between n2 and new module
            # skip any first module node that is not linked to second node
            if (AM[secondNode, firstModNode] != 0) {
                #print(c(-2, secondNode, firstModNode, AM[secondNode, firstModNode], AM[firstModNode, secondNode]))
                secondModuleDelta <- secondModuleDelta + AM[secondNode, firstModNode] + AM[firstModNode, secondNode]
            }
        }   
    }
    if (debug) {
        print(paste0("second group: ", modules[[secondModule]]))   
    }
    if (length(modules[[secondModule]]) > 0) {
        for (secondModNode in modules[[secondModule]]) {
            # find the value lost from links between n2 and its current module
            if (secondModNode != secondNode) {
                #print(c(-3, secondNode, secondModNode, -AM[secondModNode, secondNode], -AM[secondNode,secondModNode]))
                secondModuleDelta <- secondModuleDelta - AM[secondModNode, secondNode] - AM[secondNode,secondModNode]
            }
            # update for gained links between n1 and n2's module
            if (AM[firstNode, secondModNode] != 0) {
                #print(c(-4, firstNode, secondModNode,AM[firstNode, secondModNode], AM[secondModNode, firstNode]))
                firstModuleDelta <- firstModuleDelta + AM[firstNode, secondModNode] + AM[secondModNode, firstNode]
            }
        }
    }
    # calculate direction of clustering
    direction <- 0
    deltaQ <- 0 
    # favor answer if it is superior or equal
    if (firstModuleDelta >= secondModuleDelta & firstModuleDelta > 0)
    {
        deltaQ <- firstModuleDelta
        #print(c(firstModule, firstModuleDelta, secondModule, secondModuleDelta,1))
        direction <- 1
    }
    else if (secondModuleDelta > firstModuleDelta & secondModuleDelta > 0)
    {
        deltaQ <- secondModuleDelta
        direction <- -1
        #print(c(firstModule, firstModuleDelta, secondModule, secondModuleDelta,-1))
    }
    # need to access both values later
    return(c(direction,deltaQ))
}

extra <- FALSE
# actual iteration in loop 
combineModules <- function(extra) {
    
    # helps us check if we actually end up combining modules
    answer <- FALSE
    
    # move a node and test q value
    for (node in node_list$V1) {
        # check if node has outgoing edges
        if (node_list[node,3] & node_list[node,4]) {
            # identify best merging, default values indicate no changes
            #print(paste0("node: ", node))
            bestDeltaQ <- 0
            bestLink <- node
            change <- 0 
            # save current id in case of reverting
            nodeModID <- node_list[node, 2]
            
            # get all linked node_list to current node using adjacency matrix
            lN_list <- which(AM[node, ] != 0, arr.ind = TRUE)
    
            # go through all linked node_list  
            for (linkedNode in lN_list) {
                # node_list should be in different mod and inside q algorithm
                if (node_list[linkedNode, 2] != nodeModID & node_list[linkedNode, 3]) {
                    #print(paste0("linkedNode: ", linkedNode))
                    # find best value of proposed change
                    answer <- CalcShortQ(node, linkedNode)
                    direction <- answer[1]
                    deltaQ <- answer[2]
                    #print(paste0("deltaQ: ", answer[2]))
                    
                    # if the value is more advantageous 
                    if(deltaQ > bestDeltaQ) {
                        change <- direction
                        bestLink <- linkedNode
                        bestDeltaQ <- deltaQ
                        if (debug) {
                            print("more advantageous")
                            print(paste0("change: ", change))
                            print(paste0("bestDeltaQ: ", bestDeltaQ))
                        }
                    }
                }
            }
            
            # make best change in terms of modularity
            otherModID <- node_list[bestLink, 2]
            # current node goes to another module
            if(change == 1 & !(c(node) %in% modules[[otherModID]])) {
                # add node to new module 
                modules[[otherModID]][[length(modules[[otherModID]]) + 1]] <<- c(node)
                
                # change attribute of node
                node_list[node, 2] <- otherModID
             
                # remove node from current module 
                modules[[nodeModID]] <<- modules[[nodeModID]][! modules[[nodeModID]] %in% c(node)]
                # can't remove module from module list and adjacency Matrix if it is empty
                answer <- TRUE
            }
            else if(change == -1 & !(c(bestLink) %in% modules[[nodeModID]])) {
                # add node to new module 
                modules[[nodeModID]][[length(modules[[nodeModID]]) + 1]] <<- c(bestLink)
                
                # change attribute of node
                node_list[bestLink, 2] <- nodeModID
                
                # remove node from current module 
                modules[[otherModID]] <<- modules[[otherModID]][! modules[[otherModID]] %in% c(bestLink)]
                # don't remove module from module list and adjacency Matrix if it is empty 
                answer <- TRUE
            }
            if(debug) {
                print(modules)
                print(node_list)   
            }
            node_list <<- node_list
        }
    }
    
    if(extra) {
        print("still need to fill in")
    }
    
    return(answer)
}

# read in edges 
edges <- read.table(file.choose(), header = FALSE, sep = " ")
# make easily accessible in tibble 
edge_list <- tibble(from = edges[,1], to = edges[,2])

# read in node_list
node_list <- read.table(file.choose(), header = FALSE, sep = " ")
# make easily accessible in tibble 
#node_list <- data.frame(id = c(1,2,7,4,5,6,3))
node_list$moduleNumber <- c(1: nrow(node_list))
v <- rep(1, nrow(node_list))
node_list$Qstate <- as.logical(v)
node_list$isImportantCon<- as.logical(v)

# adjacency Matrix stores edges in identifiable format 
# decreases look up time for get directed link 
eg <- graph_from_data_frame(edge_list, directed = TRUE )
AM <- as_adjacency_matrix(eg,sparse = FALSE)

#modules represented by list
modules <- list(1:nrow(node_list))
for (i in modules) {
    modules[i] <- i
}

#debug option
debug = FALSE

tic("start")
mainAlgorithm()
toc()
